File: ios/MenuMakerUITests/MarketplaceFlowTests.swift
//
//  MarketplaceFlowTests.swift
//  MenuMakerUITests
//
//  Critical P0 tests for marketplace and ordering flows
//

import XCTest

final class MarketplaceFlowTests: XCTestCase {

    var app: XCUIApplication!

    override func setUpWithError() throws {
        continueAfterFailure = false
        app = XCUIApplication()
        app.launchArguments = ["UI-Testing"]
        app.launch()

        // Login first (required for marketplace access)
        let loginPage = LoginPage(app: app)
        if loginPage.emailField.waitForExistence(timeout: 2) {
            loginPage.login(email: "test@example.com", password: "password123")
            // Wait for navigation
            _ = app.tabBars.firstMatch.waitForExistence(timeout: 5)
        }

        // Navigate to marketplace tab
        let marketplaceTab = app.tabBars.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'marketplace' OR label CONTAINS[c] 'browse'")).firstMatch
        if marketplaceTab.waitForExistence(timeout: 2) {
            marketplaceTab.tap()
        }
    }

    override func tearDownWithError() throws {
        app = nil
    }

    // MARK: - Marketplace Browsing Tests (P0)

    @MainActor
    func testMarketplaceDisplaysSellers() throws {
        let marketplacePage = MarketplacePage(app: app)

        marketplacePage
            .assertScreenDisplayed()
            .assertSellersDisplayed()
            .assertSortButtonExists()
    }

    @MainActor
    func testSearchSellers() throws {
        let marketplacePage = MarketplacePage(app: app)

        marketplacePage
            .assertSellersDisplayed()
            .search("Pizza")

        // Verify search results update
        sleep(2) // Wait for search debounce
        XCTAssertTrue(marketplacePage.firstSellerCard.exists, "Search results should be displayed")
    }

    @MainActor
    func testSortByDistance() throws {
        let marketplacePage = MarketplacePage(app: app)

        marketplacePage
            .assertSellersDisplayed()
            .sortByDistance()

        // Verify sellers reordered
        sleep(1)
        XCTAssertTrue(marketplacePage.firstSellerCard.exists, "Sorted results should be displayed")
    }

    @MainActor
    func testSortByRating() throws {
        let marketplacePage = MarketplacePage(app: app)

        marketplacePage
            .assertSellersDisplayed()
            .sortByRating()

        sleep(1)
        XCTAssertTrue(marketplacePage.firstSellerCard.exists, "Sorted results should be displayed")
    }

    @MainActor
    func testFilterByCuisine() throws {
        let marketplacePage = MarketplacePage(app: app)

        marketplacePage.assertSellersDisplayed()

        // Try to select a cuisine filter (if available)
        let cuisineButton = app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'italian' OR label CONTAINS[c] 'chinese' OR label CONTAINS[c] 'indian'")).firstMatch
        if cuisineButton.waitForExistence(timeout: 2) {
            cuisineButton.tap()
            sleep(1)
            // Verify filtered results
            XCTAssertTrue(marketplacePage.firstSellerCard.exists || marketplacePage.emptyState.exists,
                         "Should show filtered results or empty state")
        } else {
            throw XCTSkip("No cuisine filters available")
        }
    }

    @MainActor
    func testPullToRefresh() throws {
        let marketplacePage = MarketplacePage(app: app)

        marketplacePage
            .assertSellersDisplayed()
            .pullToRefresh()

        sleep(2) // Wait for refresh
        marketplacePage.assertSellersDisplayed()
    }

    // MARK: - Ordering Flow Tests (P0)

    @MainActor
    func testViewSellerMenu() throws {
        let marketplacePage = MarketplacePage(app: app)
        marketplacePage.assertSellersDisplayed()

        let menuPage = marketplacePage.tapFirstSeller()
        menuPage.assertScreenDisplayed()
    }

    @MainActor
    func testAddItemToCart() throws {
        let marketplacePage = MarketplacePage(app: app)
        marketplacePage.assertSellersDisplayed()

        let menuPage = marketplacePage.tapFirstSeller()
        menuPage
            .assertScreenDisplayed()
            .addFirstItemToCart()
            .assertCartBadgeCount(1)
    }

    @MainActor
    func testViewCart() throws {
        let marketplacePage = MarketplacePage(app: app)
        marketplacePage.assertSellersDisplayed()

        // Add item to cart first
        let menuPage = marketplacePage.tapFirstSeller()
        menuPage.addFirstItemToCart()

        // Navigate to cart
        let cartPage = menuPage.navigateToCart()
        cartPage
            .assertScreenDisplayed()
            .assertCartNotEmpty()
    }

    @MainActor
    func testIncrementCartItemQuantity() throws {
        let marketplacePage = MarketplacePage(app: app)
        let menuPage = marketplacePage.assertSellersDisplayed().tapFirstSeller()
        menuPage.addFirstItemToCart()

        let cartPage = menuPage.navigateToCart()
        cartPage
            .assertCartNotEmpty()
            .incrementFirstItem()
            .assertCartNotEmpty()
    }

    @MainActor
    func testRemoveItemFromCart() throws {
        let marketplacePage = MarketplacePage(app: app)
        let menuPage = marketplacePage.assertSellersDisplayed().tapFirstSeller()
        menuPage.addFirstItemToCart()

        let cartPage = menuPage.navigateToCart()
        cartPage
            .assertCartNotEmpty()
            .removeFirstItem()

        // Should show empty state or no items
        sleep(1)
        XCTAssertTrue(cartPage.emptyCartMessage.exists || cartPage.cartItems.count == 0,
                     "Cart should be empty after removing item")
    }

    @MainActor
    func testApplyCouponCode() throws {
        let marketplacePage = MarketplacePage(app: app)
        let menuPage = marketplacePage.assertSellersDisplayed().tapFirstSeller()
        menuPage.addFirstItemToCart()

        let cartPage = menuPage.navigateToCart()
        guard cartPage.couponField.waitForExistence(timeout: 2) else {
            throw XCTSkip("Coupon feature not implemented yet")
        }

        cartPage
            .assertCartNotEmpty()
            .applyCoupon("TESTCODE")

        // Verify coupon applied or error shown
        sleep(2)
        let couponApplied = app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'discount' OR label CONTAINS[c] 'applied'")).firstMatch.exists
        let couponError = app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'invalid' OR label CONTAINS[c] 'error'")).firstMatch.exists
        XCTAssertTrue(couponApplied || couponError, "Should show coupon result")
    }

    @MainActor
    func testProceedToCheckout() throws {
        let marketplacePage = MarketplacePage(app: app)
        let menuPage = marketplacePage.assertSellersDisplayed().tapFirstSeller()
        menuPage.addFirstItemToCart()

        let cartPage = menuPage.navigateToCart()
        guard cartPage.checkoutButton.waitForExistence(timeout: 2) else {
            throw XCTSkip("Checkout button not found")
        }

        cartPage.assertCartNotEmpty()

        let checkoutPage = cartPage.proceedToCheckout()
        checkoutPage.assertScreenDisplayed()
    }

    @MainActor
    func testEmptyCartState() throws {
        // Navigate directly to cart without adding items
        let cartTab = app.tabBars.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'cart'")).firstMatch
        guard cartTab.waitForExistence(timeout: 2) else {
            throw XCTSkip("Cart tab not found")
        }
        cartTab.tap()

        let cartPage = CartPage(app: app)
        cartPage.assertScreenDisplayed()

        // Should show empty state or checkout disabled
        let isEmpty = cartPage.emptyCartMessage.exists
        let checkoutDisabled = !cartPage.checkoutButton.isEnabled

        XCTAssertTrue(isEmpty || checkoutDisabled, "Empty cart should show empty state or disable checkout")
    }
}

File: android/app/src/androidTest/kotlin/com/menumaker/MarketplaceFlowTests.kt
package com.menumaker

import androidx.compose.ui.test.junit4.createAndroidComposeRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.menumaker.pageobjects.*
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Android Tests for Marketplace and Ordering Flow
 */
@RunWith(AndroidJUnit4::class)
class MarketplaceFlowTests {

    @get:Rule
    val composeTestRule = createAndroidComposeRule<MainActivity>()

    @Before
    fun setup() {
        // Login before tests
        val loginPage = LoginPage(composeTestRule)
        Thread.sleep(2000)
        if (loginPage.emailField.exists) {
            loginPage.login("test@example.com", "password123")
            Thread.sleep(2000)
        }
    }

    // MARK: - Marketplace Browse Tests (P0)

    @Test
    fun testMarketplaceScreenDisplays() {
        val marketplacePage = MarketplacePage(composeTestRule)
        marketplacePage.assertScreenDisplayed()
    }

    @Test
    fun testSellersDisplayed() {
        val marketplacePage = MarketplacePage(composeTestRule)
        marketplacePage.assertSellersDisplayed()
    }

    @Test
    fun testSearchSellers() {
        val marketplacePage = MarketplacePage(composeTestRule)

        marketplacePage
            .search("pizza")

        Thread.sleep(1000)
        // Results should be filtered
    }

    @Test
    fun testSortByDistance() {
        val marketplacePage = MarketplacePage(composeTestRule)

        marketplacePage.sortByDistance()

        Thread.sleep(1000)
        marketplacePage.assertSellersDisplayed()
    }

    @Test
    fun testSortByRating() {
        val marketplacePage = MarketplacePage(composeTestRule)

        marketplacePage.sortByRating()

        Thread.sleep(1000)
        marketplacePage.assertSellersDisplayed()
    }

    // MARK: - Menu Viewing Tests (P0)

    @Test
    fun testViewSellerMenu() {
        val marketplacePage = MarketplacePage(composeTestRule)

        val menuPage = marketplacePage.tapFirstSeller()

        menuPage.assertScreenDisplayed()
        menuPage.assertMenuItemsDisplayed()
    }

    // MARK: - Cart Tests (P0)

    @Test
    fun testAddItemToCart() {
        val marketplacePage = MarketplacePage(composeTestRule)
        val menuPage = marketplacePage.tapFirstSeller()

        menuPage
            .addFirstItemToCart()
            .assertCartBadgeCount(1)
    }

    @Test
    fun testAddMultipleItemsToCart() {
        val marketplacePage = MarketplacePage(composeTestRule)
        val menuPage = marketplacePage.tapFirstSeller()

        menuPage
            .addFirstItemToCart()
            .addItemToCart(1)
            .assertCartBadgeCount(2)
    }

    @Test
    fun testNavigateToCart() {
        val marketplacePage = MarketplacePage(composeTestRule)
        val menuPage = marketplacePage.tapFirstSeller()

        menuPage.addFirstItemToCart()

        val cartPage = menuPage.navigateToCart()

        cartPage.assertScreenDisplayed()
        cartPage.assertCartNotEmpty()
    }

    @Test
    fun testRemoveItemFromCart() {
        val marketplacePage = MarketplacePage(composeTestRule)
        val menuPage = marketplacePage.tapFirstSeller()

        menuPage.addFirstItemToCart()

        val cartPage = menuPage.navigateToCart()

        cartPage
            .assertCartNotEmpty()
            .removeFirstItem()

        Thread.sleep(1000)
        // Cart should be empty or have fewer items
    }

    @Test
    fun testUpdateItemQuantity() {
        val marketplacePage = MarketplacePage(composeTestRule)
        val menuPage = marketplacePage.tapFirstSeller()

        menuPage.addFirstItemToCart()

        val cartPage = menuPage.navigateToCart()

        cartPage
            .assertCartNotEmpty()
            .increaseQuantity()

        Thread.sleep(500)
        // Quantity should increase
    }

    @Test
    fun testApplyCoupon() {
        val marketplacePage = MarketplacePage(composeTestRule)
        val menuPage = marketplacePage.tapFirstSeller()

        menuPage.addFirstItemToCart()

        val cartPage = menuPage.navigateToCart()

        cartPage
            .assertCartNotEmpty()
            .applyCoupon("TESTCODE")

        Thread.sleep(2000)
        // Coupon should be applied or show error
    }

    // MARK: - Checkout Tests (P0)

    @Test
    fun testProceedToCheckout() {
        val marketplacePage = MarketplacePage(composeTestRule)
        val menuPage = marketplacePage.tapFirstSeller()

        menuPage.addFirstItemToCart()

        val cartPage = menuPage.navigateToCart()
        val checkoutPage = cartPage.proceedToCheckout()

        checkoutPage.assertScreenDisplayed()
    }

    @Test
    fun testPlaceOrderWithCashPayment() {
        val marketplacePage = MarketplacePage(composeTestRule)
        val menuPage = marketplacePage.tapFirstSeller()

        menuPage.addFirstItemToCart()

        val cartPage = menuPage.navigateToCart()
        val checkoutPage = cartPage.proceedToCheckout()

        checkoutPage
            .enterDeliveryAddress("123 Test Street, Test City")
            .selectPaymentMethod(CheckoutPage.PaymentMethod.CASH)
            .placeOrder()

        Thread.sleep(3000)
        // Order should be placed successfully
    }

    @Test
    fun testPlaceOrderWithCardPayment() {
        val marketplacePage = MarketplacePage(composeTestRule)
        val menuPage = marketplacePage.tapFirstSeller()

        menuPage.addFirstItemToCart()

        val cartPage = menuPage.navigateToCart()
        val checkoutPage = cartPage.proceedToCheckout()

        checkoutPage
            .enterDeliveryAddress("123 Test Street, Test City")
            .selectPaymentMethod(CheckoutPage.PaymentMethod.CARD)

        Thread.sleep(1000)
        // Card payment form should appear
    }

    @Test
    fun testCheckoutWithEmptyAddress() {
        val marketplacePage = MarketplacePage(composeTestRule)
        val menuPage = marketplacePage.tapFirstSeller()

        menuPage.addFirstItemToCart()

        val cartPage = menuPage.navigateToCart()
        val checkoutPage = cartPage.proceedToCheckout()

        checkoutPage
            .selectPaymentMethod(CheckoutPage.PaymentMethod.CASH)
            .placeOrder()

        Thread.sleep(1000)
        // Should show validation error
    }
}
